diff --git a/config.def.h b/config.def.h
index 823e79f..b206e3a 100644
--- a/config.def.h
+++ b/config.def.h
@@ -20,6 +20,9 @@ static char *shell = "/bin/sh";
 char *utmp = NULL;
 char *stty_args = "stty raw pass8 nl -echo -iexten -cstopb 38400";
 
+/* popenscreen */
+static char *popenscreencmd = "tr ' ' '\n' | sort | uniq | dmenu -b -l 5 -w $WINDOW -fn 'Gohufont:pixelsize=14' -nb '#ffffff' -nf '#000000' -sb '#000000' -sf '#ffffff' | tr '\n' ' '";
+
 /* identification sequence returned in DA and DECID */
 char *vtiden = "\033[?6c";
 
@@ -166,6 +169,7 @@ static MouseShortcut mshortcuts[] = {
 
 static Shortcut shortcuts[] = {
 	/* mask                 keysym          function        argument */
+	{ TERMMOD,              XK_P,           popenscreen,    {.i =  0} },
 	{ XK_ANY_MOD,           XK_Break,       sendbreak,      {.i =  0} },
 	{ ControlMask,          XK_Print,       toggleprinter,  {.i =  0} },
 	{ ShiftMask,            XK_Print,       printscreen,    {.i =  0} },
diff --git a/st.c b/st.c
index 46cf2da..8e04626 100644
--- a/st.c
+++ b/st.c
@@ -223,7 +223,7 @@ static Term term;
 static Selection sel;
 static CSIEscape csiescseq;
 static STREscape strescseq;
-static int iofd = 1;
+int iofd = 1;
 static int cmdfd;
 static pid_t pid;
 
diff --git a/x.c b/x.c
index 00cb6b1..c1f1b86 100644
--- a/x.c
+++ b/x.c
@@ -52,6 +52,7 @@ typedef struct {
 static void clipcopy(const Arg *);
 static void clippaste(const Arg *);
 static void numlock(const Arg *);
+static void popenscreen(const Arg *);
 static void selpaste(const Arg *);
 static void zoom(const Arg *);
 static void zoomabs(const Arg *);
@@ -240,6 +241,10 @@ static char *opt_title = NULL;
 
 static int oldbutton = 3; /* button event on startup: 3 = release */
 
+/* popenscreen externs */
+extern int iofd;
+extern char *popenscreencmd;
+
 void
 clipcopy(const Arg *dummy)
 {
@@ -308,6 +313,55 @@ zoomreset(const Arg *arg)
 	}
 }
 
+void
+popenscreen(const Arg *dummy)
+{
+	int r[2], w[2];
+
+	pipe(r);
+	pipe(w);
+
+	int pid = fork();
+	switch (pid) {
+	case -1:
+		return;
+	case 0:
+		close(iofd);
+		setsid(); /* create a new process group */
+
+		dup2(r[0], STDIN_FILENO);
+		dup2(w[1], STDOUT_FILENO);
+
+		close(r[1]);
+		close(w[0]);
+
+		char win[16];
+		snprintf(win, sizeof(win), "%lu", xw.win);
+		setenv("WINDOW", win, 1);
+
+		execlp("/bin/sh", "sh", "-c", popenscreencmd, NULL);
+		die("execlp:");
+	}
+
+	signal(SIGCHLD, SIG_IGN);
+
+	close(r[0]);
+	close(w[1]);
+	int oldiofd = iofd;
+	iofd = r[1];
+	printscreen(NULL);
+	iofd = oldiofd;
+
+	close(r[1]);
+	int n;
+	char buf[256];
+
+	while ((n = read(w[0], &buf, sizeof(buf) - 1)) > 0) {
+		buf[n] = '\0';
+		ttywrite(buf, n, 1);
+	}
+}
+
 int
 evcol(XEvent *e)
 {
